frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -3258,7 +3258,8 @@ const IR::Node* TypeInference::postorder(IR::KeyElement* elem) {
     while (ktype->is<IR::Type_Newtype>())
         ktype = ktype->to<IR::Type_Newtype>()->type;
     if (!ktype->is<IR::Type_Bits>() && !ktype->is<IR::Type_Enum>() &&
-        !ktype->is<IR::Type_Error>() && !ktype->is<IR::Type_Boolean>())
+        !ktype->is<IR::Type_Error>() && !ktype->is<IR::Type_Boolean>() &&
+        !ktype->is<IR::Type_Name>() && !ktype->is<IR::Type_SerEnum>())
```

Comment:
this can never be Type_Name.

Reply:
Testing shows that it does.  See https://github.com/p4lang/p4c/issues/1739
When the above code change was applied the issue is fixed  - the frontend does not block `type`.

Reply:
You should use getType(ktype) instead of ktype.

Reply:
Not quite.  The nesting combines typedef and `type`.  Both types have Type_Name common between them.  To make less code changes, adding a test for `IR::Type_Name` is the smallest and correct change.  Otherwise, I also have to add a while loop for typdef in this method akin to the while loop for Type_Newtype.  Additionally, typeChecker has a separate method to verify, for example, a header does not gets allowed in a `type`: https://github.com/p4lang/p4c/blob/master/frontends/p4/typeChecking/typeChecker.cpp#L1289


Please see `testdata/p4_16_samples/p416-type-use3.p4` used to test. 

I removed the `!ktype->is<IR::Type_Name>()` check and get these errors below. See 'Path name` in the log below to see what objects failed.  How is getType(ktype) used in such failure situation to fix the issue?


```
../testdata/p4_16_samples/p416-type-use3.p4(159): [--Werror=type-error] error: Key hdr.custom.eddt field type must be a scalar type; it cannot be CustomDD_t
            hdr.custom.eddt : exact;
            ^^^^^^^^^^^^^^^
[1130] Type_Name
  path: [1129] Path name=CustomDD_t absolute=0
../testdata/p4_16_samples/p416-type-use3.p4(160): [--Werror=type-error] error: Key hdr.custom.edtd field type must be a scalar type; it cannot be CustomD_t
            hdr.custom.edtd : exact;
            ^^^^^^^^^^^^^^^
[1110] Type_Name
  path: [1109] Path name=CustomD_t absolute=0
../testdata/p4_16_samples/p416-type-use3.p4(161): [--Werror=type-error] error: Key hdr.custom.edtt field type must be a scalar type; it cannot be CustomDT_t
            hdr.custom.edtt : exact;
            ^^^^^^^^^^^^^^^
[1140] Type_Name
  path: [1139] Path name=CustomDT_t absolute=0
../testdata/p4_16_samples/p416-type-use3.p4(163): [--Werror=type-error] error: Key hdr.custom.etdt field type must be a scalar type; it cannot be CustomTD_t
            hdr.custom.etdt : exact;
            ^^^^^^^^^^^^^^^
[1150] Type_Name
  path: [1149] Path name=CustomTD_t absolute=0
../testdata/p4_16_samples/p416-type-use3.p4(164): [--Werror=type-error] error: Key hdr.custom.ettd field type must be a scalar type; it cannot be CustomT_t
            hdr.custom.ettd : exact;
            ^^^^^^^^^^^^^^^
[1120] Type_Name
  path: [1119] Path name=CustomT_t absolute=0
../testdata/p4_16_samples/p416-type-use3.p4(165): [--Werror=type-error] error: Key hdr.custom.ettt field type must be a scalar type; it cannot be CustomTT_t
            hdr.custom.ettt : exact;
            ^^^^^^^^^^^^^^^
[1160] Type_Name
  path: [1159] Path name=CustomTT_t absolute=0
```

Reply:
Use the model of the minWidthInBits function from the typeMap from the upcoming PR to resolve the types. Note that it never has to worry about type_name, and it does work for any nesting depth.

Reply:
Note minWidthInBits does not support Type_Error and Type_Enum.  So, I made the following change to typeChecker. 
```
@@ -3255,11 +3326,8 @@ const IR::Node* TypeInference::postorder(IR::KeyElement* elem) {
     auto ktype = getType(elem->expression);
     if (ktype == nullptr)
         return elem;
-    while (ktype->is<IR::Type_Newtype>())
-        ktype = ktype->to<IR::Type_Newtype>()->type;
-    if (!ktype->is<IR::Type_Bits>() && !ktype->is<IR::Type_Enum>() &&
-        !ktype->is<IR::Type_Error>() && !ktype->is<IR::Type_Boolean>() &&
-        !ktype->is<IR::Type_Name>() && !ktype->is<IR::Type_SerEnum>())
+    if (!ktype->is<IR::Type_Enum>() && !ktype->is<IR::Type_Error>() &&
+        !(typeMap->keyElementValid(ktype, elem->expression) > 0))
         typeError("Key %1% field type must be a scalar type; it cannot be %2%",
                   elem->expression, ktype->toString());
     auto type = getType(elem->matchType);
```

The new function in keyElementValid() is an abbreviated minWidthInBits.

```
diff --git a/frontends/p4/typeMap.cpp b/frontends/p4/typeMap.cpp
+int TypeMap::keyElementValid(const IR::Type* type, const IR::Node* errorPosition) {
+    auto t = getTypeType(type, true);
+    if (auto tb = t->to<IR::Type_Bits>()) {
+        return tb->width_bits();
+    } else if (auto te = t->to<IR::Type_SerEnum>()) {
+        return keyElementValid(te->type, errorPosition);
+    } else if (t->is<IR::Type_Boolean>()) {
+        return 1;
+    } else if (auto tnt = t->to<IR::Type_Newtype>()) {
+        return keyElementValid(tnt->type, errorPosition);
+    }
+
+    ::error("%1%: width not well-defined", errorPosition);
+    return -1;
+}
 
 }  // namespace P4
diff --git a/frontends/p4/typeMap.h b/frontends/p4/typeMap.h
index c3b0c8e9..a8c54f71 100644
--- a/frontends/p4/typeMap.h
+++ b/frontends/p4/typeMap.h
@@ -91,6 +91,11 @@ class TypeMap final : public ProgramMap {
 
     // Used for tuples and stacks only
     const IR::Type* getCanonical(const IR::Type* type);
+    /// The minimum width in bits of this type.  If the width is not
+    /// well-defined this will report an error and return -1.
+    int minWidthBits(const IR::Type* type, const IR::Node* errorPosition);
+    /// Used by typeChecker to validate keyElement.
+    int keyElementValid(const IR::Type* type, const IR::Node* errorPosition);
 };
 }  // namespace P4
```

Reply:
I may be misunderstanding your comment, so I will ask: Do you think there is a reason that minWidthInBits _should_ work for type `error` or (non-serializable) `enum`?  It seems to me that it should not, if its intent is to return a target-independent width in bits for a type.  There is no such target-independent value for those types.

Reply:
I have used a simple algorithm.  Retain the check for types which are not supported by minWidthInBits and remove checks for types that are.  Then, abbreviate minWidthInBits to not support any type which doesn't exist in the original check in typeChecker, e.g., Varbits.

Yes, minWidthInBits should not work for type `error` nor `enum`.